onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,401 @@
+        softmaxOperatorDesc.AxisCount = gsl::narrow_cast<uint32_t>(axes.size());
+        softmaxOperatorDesc.Axes = axes.data();
+        const DML_OPERATOR_DESC softmaxDesc = {DML_OPERATOR_ACTIVATION_SOFTMAX1, &softmaxOperatorDesc};
+
+        uint32_t reshapedTransposedOutputTensorShape[4] {batchSize, numHeads, sequenceLength, headSize};
+        uint32_t reshapedTransposedOutputTensorStride[4] {sequenceLength * numHeads * headSize, headSize * sequenceLength, headSize, 1};
+        TensorDesc reshapedTransposedOutputTensorDesc = TensorDesc(
+                                                            GetDmlDataTypeFromMlDataType(dataType),
+                                                            reshapedTransposedOutputTensorShape,
+                                                            reshapedTransposedOutputTensorStride,
+                                                            0 // guaranteedBaseOffsetAlignment
+                                                        );
+        DML_TENSOR_DESC namedReshapedTransposedOutputTensorDesc = reshapedTransposedOutputTensorDesc.GetDmlDesc();
+
+        DML_GEMM_OPERATOR_DESC attentionWeightOperatorDesc = {};
+        attentionWeightOperatorDesc.ATensor = &namedQueryKeyTensorDesc;
+        attentionWeightOperatorDesc.BTensor = &namedReshapedTransposedQueryTensorDesc;
+        attentionWeightOperatorDesc.CTensor = nullptr;
+        attentionWeightOperatorDesc.OutputTensor = &namedReshapedTransposedOutputTensorDesc;
+        attentionWeightOperatorDesc.TransA = DML_MATRIX_TRANSFORM_NONE;
+        attentionWeightOperatorDesc.TransB = DML_MATRIX_TRANSFORM_NONE;
+        attentionWeightOperatorDesc.Alpha = 1.0f;
+        attentionWeightOperatorDesc.Beta = 0.0f;
+        attentionWeightOperatorDesc.FusedActivation = nullptr;
+        const DML_OPERATOR_DESC attentionWeightDesc {DML_OPERATOR_GEMM, &attentionWeightOperatorDesc};
+
+        TensorDesc transposedOutputTensorDesc = TensorDesc(
+                                                    m_outputTensorDescs[0].GetDmlDataType(),
+                                                    m_outputTensorDescs[0].GetSizes(),
+                                                    std::nullopt,
+                                                    0 // guaranteedBaseOffsetAlignment
+                                                );
+        DML_TENSOR_DESC namedTransposedOutputTensorDesc = transposedOutputTensorDesc.GetDmlDesc();
+
+        DML_ACTIVATION_LINEAR_OPERATOR_DESC outputOperatorDesc = {};
+        outputOperatorDesc.Alpha = 1.0f;
+        outputOperatorDesc.Beta = 0.0f;
+        outputOperatorDesc.InputTensor = &outputDescs[0];
+        outputOperatorDesc.OutputTensor = &namedTransposedOutputTensorDesc;
+        const DML_OPERATOR_DESC outputDesc {DML_OPERATOR_ACTIVATION_LINEAR, &outputOperatorDesc};
+
+
+        MLOperatorGraphDesc operatorGraphDesc = {};
+        operatorGraphDesc.nodeCount = 10;
```

Comment:
>operatorGraphDesc.nodeCount = 10; [](http://example.com/codeflow?start=8&length=33)

`operatorGraphDesc.nodeCount = uint32_t(opDescs.size());`

Less brittle. Note I didn't bother with the narrow_cast here because it's not a dynamic array.

Reply:
The enum approach works well too 👍.